File: Quick-Sort/js/leetcode_lc_169.js
const nums = [2,2,1,1,1,2,2]

/**
 * @param {number[]} nums
 * @return {number}
 */
var majorityElement = function(nums) {
    quicksort(nums,0,nums.length-1)
    return nums[Math.floor(nums.length/2)]
};

function quicksort(arr,start,end){
    if(start >= end){
        return 
    }

    const pivotIdx = partition(arr,start,end)

    quicksort(arr,start,pivotIdx-1)
    quicksort(arr,pivotIdx+1,end)
}

function partition(arr,start,end){
    let leftIdx = start , rightIdx = end
    const pivot = arr[start]

    while(leftIdx < rightIdx){
        while(arr[rightIdx] > pivot && leftIdx < rightIdx){
            rightIdx--
        }

        while(arr[leftIdx] <= pivot && leftIdx < rightIdx){
            leftIdx++
        }

        if(leftIdx < rightIdx){
            [arr[leftIdx],arr[rightIdx]] = [arr[rightIdx],arr[leftIdx]]
        }
    }

    arr[start] = arr[leftIdx]
    arr[leftIdx] = pivot
    return leftIdx
}


console.log(`返回${nums}的多数元素`,majorityElement(nums))

File: Quick-Sort/python/leetcode_lc_169.py
nums = [3,2,3]

class Solution:
    def majorityElement(self, nums):

        self.quickSort(nums,0,len(nums)-1)

        return nums[int(len(nums)/2)]

    def quickSort(self,arr,start,end):
        if start >= end:
            return 
        
        leftIdx,rightIdx = self.partition(arr,start,end)

        self.quickSort(arr,start,leftIdx-1)
        self.quickSort(arr,rightIdx,end)

    def partition(self,arr,start,end):
        i = left = start
        right = end + 1

        pivot = arr[start]

        while i < right :
            if arr[i] < pivot:
                arr[left+1],arr[i] = arr[i],arr[left+1]
                i+=1
                left+=1
            elif arr[i] > pivot:
                arr[i],arr[right-1] = arr[right-1],arr[i]
                right-=1
            else:
                i+=1

        arr[start] = arr[left]
        arr[left] = pivot

        return left,right


solution = Solution()

print("数组{} 其中的多数元素为 {} ".format(nums,solution.majorityElement(nums)))